### ISSUE TITLE
Onedata integration with the provided identity federations (eduGAIN and Elixir AAI) via KeyCloack

### ISSUE THREAD

**Issue by cgizzi** [rhea-group/hn-prototype-platform#2]
The federated identities must also be available when working with the data management services. Onedata is already capable of using external identity providers that use the OpenID Connect (OIDC) protocol for authentication. This is the native KeyCloak protocol for relying services, so integration should be straightforward. Effort will be required to verify the correct integration with the identity federations.

**Comment by bkryza** [comment]
Onedata Onezone service supports external OIDC providers, which can be configured using [auth.config file](https://onedata.org/#/home/documentation/doc/administering_onedata/openid_configuration.html), which needs to be configured before starting Onezone service. From then on, users can login and use Onedata services, including data management, using their original identity.

Currently, we do not have a specific plugin for Keycloak, however as most OIDC plugins follow the same pattern, it could be possible to use one of the existing configuration entries (e.g. for Google or Dropbox) by replacing relevant entries with Keycloak endpoints. 

Example configuratoin entry for Google OIDC is as follows:

```erlang
    {google, [
        % Standard config
        {auth_module, auth_google},
        {app_id, <<"APP_ID">>},
        {app_secret, <<"APP_SECRET">>},
        % Provider specific config
        {xrds_endpoint,
            <<"https://accounts.google.com/.well-known/openid-configuration">>}
    ]},
```

@konstan @loomis Please let me know if this sounds reasonable?

**Comment by mebster** [comment]
@konstan @loomis ?

**Comment by konstan** [comment]
Waiting for 3.0.0-rc15 of Onedata to be released.  Here is the procedure provided by Onedata on 17th May

```
1) You would need the newest version of OneZone that includes OIDC handler for keycloak. Currently it is being integrated into our develop stream, and after testing will be released. As far as i know, the 3.0.0-rc15 version will be released within days and should contain this feature.

2) All you need to do is edit the "auth.config" file and add an entry for RHEA KeyCloak. Refer to documentation for its location and more info - https://onedata.org/#/home/documentation/doc/administering_onedata/openid_configuration.html. The RHEA entry is similar to other entries and looks like this:

{rhea, [
        % Standard config
        {auth_module, auth_keycloak},
        {app_id, <<"OnezoneTest">>},
        {app_secret, <<"secretly-secret">>},
        % Provider specific config
        {xrds_endpoint, <<"https://fed-id.nuv.la/auth/realms/onedata/.well-known/openid-configuration">>}
    ]}

Remember that OneZone restart is required to reload the config. After this is added, a login button for RHEA KeyCloak should appear on your login screen.
```

**Comment by konstan** [comment]
Onedata themselves already tested this internally.

**Comment by cjdcordeiro** [comment]
on the Keycloak side, the application client has been tested by OneData on a separate Test Realm. Once @konstan can replicate the above procedure and validate it works as expected, we'll move this Keycloak client to production

**Comment by bkryza** [comment]
@konstan 

Below is an example auth.config entry for KeyCloak:

```erlang
    {rhea, [
        % Standard config
        {auth_module, auth_keycloak},
        {app_id, <<"APP_ID">>},
        {app_secret, <<"APP_SECRET">>},
        % Provider specific config
        {xrds_endpoint, <<"https://fed-id.nuv.la/auth/realms/onedata/.well-known/openid-configuration">>}
    ]},
```

**Comment by konstan** [comment]
@bkryza 

I added Keycloak config to auth.config as suggested above and restarted oz_worker.  This is the result.

![screen shot 2017-06-07 at 14 14 38](https://user-images.githubusercontent.com/353767/26878348-78495f2c-4b8d-11e7-966f-f460ce5f5b3a.png)


But after clicking on the sixsq button I'm getting

<img width="769" alt="screen shot 2017-06-07 at 14 28 13" src="https://user-images.githubusercontent.com/353767/26878401-9a107dd4-4b8d-11e7-8a80-c7ef9bcd70ce.png">

with the following in the logs

```
==> error.log <==
2017-06-07 12:08:31.601 [error] <0.10265.0>@gui_ws_handler:handle_RPC_req:481 Error while handling GUI RPC request - error:undef
Stacktrace:
    lists:foreach/2 line 1338
    gui_ws_handler:'-process_requests_async/1-fun-0-'/1 line 584
    gui_ws_handler:handle_decoded_message/1 line 331
    gui_ws_handler:handle_RPC_req/1 line 468
    public_rpc_backend:handle/2 line 73
    auth_keycloak:get_redirect_url(false)
```

For auth.config I re-used your OnezoneTest in https://fed-id.nuv.la where I changed `Root URL` and  `Valid Redirect URIs`, and added `Base URL`.

![screen shot 2017-06-07 at 14 14 15](https://user-images.githubusercontent.com/353767/26878467-db2510b4-4b8d-11e7-8830-c712d45f8a22.png)

```
[
    {basicAuth, []},

    {rhea, [
        % Standard config
        {auth_module, auth_keycloak},
        {app_id, <<"OnezoneTest">>},
        {app_secret, <<"8712ed65- ... -270a30291e76">>},
        % Provider specific config
        {xrds_endpoint, <<"https://fed-id.nuv.la/auth/realms/onedata/.well-known/openid-configuration">>}
     ]}
```

Can you please help with this?  Thanks.

**Comment by bkryza** [comment]
@konstan Sorry, my colleague just checked and it seems this feature commit was not included in the release branch for rc15. We will create a public docker containers with this included today and I'll let you know it's ID for testing...

**Comment by konstan** [comment]
@bkryza  Hello.  Is there any progress on this?  Thanks.

**Comment by bkryza** [comment]
@konstan 

Yes, this is the Onezone container with included KeyCloak hotfix:

```
onedata/onezone:ID-0b5c927a1d
```

Sorry again for this...

**Comment by loomis** [comment]
@konstan @cjdcordeiro Can you provide a short summary of the current testing of Keycloak/Onedata integration?

**Comment by loomis** [comment]
@cjdcordeiro What is the current status concerning the mapping between BG tenants and realms?

**Comment by cjdcordeiro** [comment]
@konstan this is how you register new clients from the command line:
```bash
curl -X POST \
    -d '{ "clientId": "clientName" }' \
    -H "Content-Type:application/json" \
    -H "Authorization: bearer $initialTokenX$" \
    https://fed-id.nuv.la/auth/realms/$REALM_NAMR$/clients-registrations/default
```

To get an initialTokenX, you login into Keycloak and go to: Realm Settings -> Client Registration -> Initial Access Tokens -> Create

**Comment by cjdcordeiro** [comment]
> @cjdcordeiro What is the current status concerning the mapping between BG tenants and realms?

waiting on final appraisal that we should do that and investigating in the meanwhile if there is a way we can copy realms to avoid doing configuration replicas by hand

**Comment by konstan** [comment]
> @konstan @cjdcordeiro Can you provide a short summary of the current testing of Keycloak/Onedata integration?

I used this image onedata/onezone:ID-0b5c927a1d as provided by @bkryza. After configuration, Onezone login interface shows the new button for authn via the configured Keycloak instance.  Login works.  However, the name of the logged in user is UNKNOWN.  I used the same realm/client (onedata/OnezoneTest) as it was used initially by Onedata to test the implementation. 
I'm re-deploying to validate this again and see if we can understand what attributes sent by Keycloak are misinterpreted/missing/etc.
More news a bit later.

**Comment by konstan** [comment]
Update after debugging this Łukasz Opioła.

- (at the moment) Onezone expects - `sub`, `email`, `login`, `name` - attributes from Keycloak.  `name`, as sent by Keycloak, is a concatination of First Name and Last Name.

![screen shot 2017-06-09 at 11 18 34](https://user-images.githubusercontent.com/353767/26969272-73358604-4d05-11e7-90d2-e644d431fc68.png)

<img width="1680" alt="screen shot 2017-06-09 at 11 18 01" src="https://user-images.githubusercontent.com/353767/26969287-83898e38-4d05-11e7-8bb8-5290122b22ee.png">

So, the UNKNOWN was because for that `konstan` user we din't have any attributes defined.  When dealing with users on the Keycloak side, we have to have a set of user mandatory attributes that satisfy Onezone / SlipStream / etc. services.

Onezone shouldn't authn / create a user if the mandatory attributes are missing.

User's info received from Keycloak is written ones in the Onezone's DB and not updated.  IMO, Onezone should be updating the user info, provided user identification is made on user's ID coming from Fed-IdP (eb2e99fd-9b4f-4225-ba6c-4a8e81cc08d1 in the above screenshot).  This would allow for users to have their personal info up-to-date (e.g. email address) in all the systems.

- bugs found in Onezone:
  - the authn configuration section for `auth.config` is hardcoded in Onezone and should be called `rhea`.  However, if it is set to some other value `oz_worker` starts OK, Onezone web UI "shows" a hidden link for the new authn option, after clicking on the hidden link - we get Server Error popup.
  - Onezone `admin` user can't delete users in Onezone via API.
  - `admin` privileges reported by API are empty.

Next steps.
- test Onezone authn via Fed-IdP with 
  - [ ] eduGAIN
  - [ ] ELIXIR

**Comment by konstan** [comment]
eduGAIN authn works.  However, the user is given the name that is taken from email address.

**Comment by luman75** [comment]
@konstan what you mean is taken fro email address?  john.doe@example.com will lead to user name jhon.doe ?

But the most important is if we are getting user email in the attributes and the user ID from IdP?

**Comment by konstan** [comment]
@luman75 

Actually, I might have been wrong.  My username in unitedid.org is konstan and email konstan@sixsq.com.  

<img width="338" alt="screen shot 2017-06-09 at 18 47 25" src="https://user-images.githubusercontent.com/353767/26985566-4e86ac22-4d44-11e7-94e3-1453e2a4cd55.png">

The name is Jay Random, which I was expecting to appear in the right upper corner after authn with Onezone, but it was KONSTAN instead.  On the left, in Authn Settings the email was konstan@sixsq.com.

Now we probably have to check with Łukasz Opioła and also see what set of attributes unitedid.org returns to Onezone.  We will do that on Monday.

**Comment by konstan** [comment]
Just tried again via https://fed-id.nuv.la -> eduGAIN -> united ID.  Here is the result.

<img width="1043" alt="screen shot 2017-06-09 at 18 57 12" src="https://user-images.githubusercontent.com/353767/26985822-7c589ff6-4d45-11e7-8e7c-065227a84b40.png">

**Comment by lopiola** [comment]
We can look into that next week. Anyway, on our side we try all possible ways to learn user name, checking the JSON attributes in the following order:
name -> login -> email before '@' -> "Unknown Name" (if none was specified)

In this case we might have taken login or email, hard to say. You can check if the account in unitedid was properly mapped to your keycloak user data, i.e. if first and last name were passed.

**Comment by lopiola** [comment]
Concerning the bugs identified a couple of comments ago:
1) the authn configuration section for auth.config is hardcoded in Onezone and should be called rhea. However, if it is set to some other value oz_worker starts OK, Onezone web UI "shows" a hidden link for the new authn option, after clicking on the hidden link - we get Server Error popup.
2) Onezone admin user can't delete users in Onezone via API.
3) admin privileges reported by API are empty.

Number 1) is to be done, we have a ticket for that and will look into it. I guess we want to do it along with couple of features enriching OpenID configuration, e.g. specifying required fields, so it might take a while.

2) Number two was a small bug and is already fixed, will be included in newest release and I have also generated a new docker with the hotfix -> onedata/onezone:ID-0b5c927a1d-2

3) Number three is invalid, which I realized only after we finished the debugging. The admin user itself does not have any privileges, so the REST response was correct, instead he is automatically added to admins groups, which does have the privileges. To view the effective (inherited) privileges of the user, you can use the endpoint:
```
/user/effective_privileges
```
which works as expected.

**Comment by loomis** [comment]
@lopiola @bkryza Yesterday, we reviewed the information that we're getting from eduGAIN and what could be used as a unique user identifier for both Nuvla and Onedata. In the end, the only value deemed suitably unique and "user friendly" was the email address of the user.  We also verified that roles and groups defined within Keycloak can be passed through the JWT token to Nuvla and Onedata.  To be precise, the attributes that will appear in the Keycloak JWT token are:

 - email: To be used as the unique identifier for the user. 
 - realm: Contains the name of the realm that was used to authenticate the user. This will be a unique string for each Buyers Group organization.
 - roles: Contains a possibly empty list of roles that the user possesses.  Each value (role) is a simple string.
 - groups: Contains a possibly empty list of groups to which the user belongs. Each value (group) is a string denoting a slash-separated path.

There are a number of other attributes that will also be passed through the JWT token, but do not have an impact on the authorization decisions.  These (e.g. given name, surname) may be useful for automatic creation of accounts or for the UI, however. @cjdcordeiro can provide the full list of expected attributes.

For authorization decisions and for identifiers used within ACLs:

 - Any token that does not have the email address is invalid and the user must be treated as unauthenticated/anonymous.
 - For the user identifier, the full email address is to be used. All authorization decisions must use the full email address.  The ACLs must also use the full email address to tie rights to a given user.
 - To ensure that roles are globally unique, the **effective role** used in authorization decisions and ACLs, must be the tuple [realm, role].  This can be represented in the ACLs in anyway that is convenient for users. For Nuvla we will represent this as a prefixed string "realm:role".
 - The groups within Keycloak are **hierarchical**, so a group like "/left/center/right" also implies membership in "/left" and "/left/center". Only the most specific groups will appear explicitly in the "groups" attributes.
 - As for roles, groups must also be globally unique.  For authorization decisions and ACLs, the **effective group** must be the tuple [realm, group].  This can be represented in the ACLs in anyway that is convenient for users.  For Nuvla, we will represent this as a prefixed string "realm:/left/center".

We will be making the changes on the Nuvla side to implement this in the next couple of days.  The changes will appear in the release we'll create on Friday.  If you see any problems with this proposal or with implementing it on the Onedata side, please let us know quickly.

@cjdcordeiro @konstan @elegoff Please correct any mistakes that I've made in the above summary of yesterday's discussions.

**Comment by lopiola** [comment]
I'd like to make sure that we understood the idea of user id correctly.

You are talking about unique and user friendly user id. Why would user id be user friendly? The idea is that user id is unique, non-changing identifier which will always be mapped to the same Onedata user on our side. There is no need for it to be "user friendly", as users do not typically see their id. Can you comment on that?

Anyway, if you can ensure that emails won't change (which I find doubtful), you can just use the "sub" field in JWT and put the email address there, for us it will just be an opaque user id. But we are not happy with using the "email" field as user id, because this is non standard.

The other thing is displayed user name, for which we can use the email address if you wish. Then, ACLs and other authorization mechanisms will use the email as desired.

**Comment by tkotisis** [comment]
@loomis 

> email: To be used as the unique identifier for the user.

In this case, would users with the same email authenticating from different IdPs be mapped to the same account?

Please note test case [TA01-008](https://github.com/SixSq/hn-prototype-platform/blob/master/test/Test%20Case%20Specification.md#ta01-008-edugain-aai---authentication-attempt-from-multiple-edugain-idps-with-credentials-sharing-a-common-email-address)

**Comment by loomis** [comment]
@lopiola 

> You are talking about unique and user friendly user id. Why would user id be user friendly? The idea is that user id is unique, non-changing identifier which will always be mapped to the same Onedata user on our side. There is no need for it to be "user friendly", as users do not typically see their id. Can you comment on that?

The user ids will appear in the ACLs and need to be used by data/account manager to allow access to resources.  Users will also need to be able to know their identifier so that they can communicate this to the data/account managers.  These identifiers are visible in the Nuvla ACLs and can be manipulated directly. Because of this, the user identifier ideally would be both fairly short and intuitive.

> Anyway, if you can ensure that emails won't change (which I find doubtful), you can just use the "sub" field in JWT and put the email address there, for us it will just be an opaque user id. But we are not happy with using the "email" field as user id, because this is non standard.

Sure, we can use the `sub` field for the user identifier.  We'll see if we settle on the email address or something else for the value.  In any case, the access control must use the value that is given here.

**Comment by loomis** [comment]
@tkotisis 

> In this case, would users with the same email authenticating from different IdPs be mapped to the same account?

Yes, they would be mapped to the same account.  Where is the BG's documents did you get the requirements supporting TA01-008?  I've don't remember seeing that requirement.

**Comment by luman75** [comment]
@loomis @lopiola based on our past experience user are getting confused when while using two or more different IdPs their accounts are the same.  Simply I'm expecting to be a diffrent user while login tough G+ or eguGain despite the fact in both systems I'm using the same email.

**Comment by tkotisis** [comment]
@loomis 
It's in the use cases provided by the BG.

In particular use case 2.1 (Edugain AAI).

**Comment by loomis** [comment]
@tkotisis Thanks.  I missed that requirement when reading through the use cases.  In this case, we'll have to use something other than the email address.  In any case, the value we choose will be put into the `sub` key of the token.  The other fields will be as specified above.
